### ISSUE TITLE
有很多个预测框

### ISSUE THREAD

**Issue by QMLmulan** [bubbliiiing/detr-pytorch#20]
导儿~我用你的DETR代码训练，预测结果产生同一个目标出现很多个预测框，需要如何解决呀？这个是端到端的，不用nms，照道理应该不会这样吧

**Comment by 7W7W7W** [comment]
> `可以试试只保留置信度最高的一个框，utils_bbox.py文件中修改为：

`@torch.no_grad()
    def forward(self, outputs, target_sizes, confidence):
        out_logits, out_bbox = outputs['pred_logits'], outputs['pred_boxes']

        assert len(out_logits) == len(target_sizes)
        assert target_sizes.shape[1] == 2

        prob = F.softmax(out_logits, -1)
        scores, labels = prob[..., :-1].max(-1)

        # 转换为 [x0, y0, x1, y1] 格式
        boxes = self.box_cxcywh_to_xyxy(out_bbox)
        img_h, img_w = target_sizes.unbind(1)
        img_h = img_h.float()
        img_w = img_w.float()
        scale_fct = torch.stack([img_w, img_h, img_w, img_h], dim=1)
        boxes = boxes * scale_fct[:, None, :]

        results = []
        for b_idx in range(len(boxes)):
            box_list = boxes[b_idx]
            score_list = scores[b_idx]
            label_list = labels[b_idx]

            unique_labels = torch.unique(label_list)  # 获取所有类别
            filtered_boxes = []

            for cls in unique_labels:
                cls_mask = (label_list == cls) & (score_list > confidence)
                cls_boxes = box_list[cls_mask]
                cls_scores = score_list[cls_mask]

                if len(cls_boxes) > 0:
                    # 选择该类别中置信度最高的框
                    max_score_idx = torch.argmax(cls_scores)
                    filtered_boxes.append(torch.cat([
                        cls_boxes[max_score_idx].unsqueeze(0),  # 边界框
                        cls_scores[max_score_idx].unsqueeze(0).unsqueeze(0),  # 置信度
                        cls.unsqueeze(0).unsqueeze(0).float(),  # 类别标签
                    ], dim=-1))

            if len(filtered_boxes) > 0:
                results.append(torch.cat(filtered_boxes, dim=0))
            else:
                results.append([])

        return results``

**Comment by QMLmulan** [comment]
已收到您的邮件，我会尽快给你回复。谢谢

**Comment by QMLmulan** [comment]
谢谢这位朋友




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2025年2月28日(星期五) 下午5:44
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [bubbliiiing/detr-pytorch] 有很多个预测框 (Issue #20)












   
可以试试只保留置信度最高的一个框，utils_bbox.py文件中修改为：
 ***@***.***_grad()
 def forward(self, outputs, target_sizes, confidence):
 out_logits, out_bbox = outputs['pred_logits'], outputs['pred_boxes']
     assert len(out_logits) == len(target_sizes)     assert target_sizes.shape[1] == 2     prob = F.softmax(out_logits, -1)     scores, labels = prob[..., :-1].max(-1)     # 转换为 [x0, y0, x1, y1] 格式     boxes = self.box_cxcywh_to_xyxy(out_bbox)     img_h, img_w = target_sizes.unbind(1)     img_h = img_h.float()     img_w = img_w.float()     scale_fct = torch.stack([img_w, img_h, img_w, img_h], dim=1)     boxes = boxes * scale_fct[:, None, :]     results = []     for b_idx in range(len(boxes)):         box_list = boxes[b_idx]         score_list = scores[b_idx]         label_list = labels[b_idx]         unique_labels = torch.unique(label_list)  # 获取所有类别         filtered_boxes = []         for cls in unique_labels:             cls_mask = (label_list == cls) &amp; (score_list &gt; confidence)             cls_boxes = box_list[cls_mask]             cls_scores = score_list[cls_mask]             if len(cls_boxes) &gt; 0:                 # 选择该类别中置信度最高的框                 max_score_idx = torch.argmax(cls_scores)                 filtered_boxes.append(torch.cat([                     cls_boxes[max_score_idx].unsqueeze(0),  # 边界框                     cls_scores[max_score_idx].unsqueeze(0).unsqueeze(0),  # 置信度                     cls.unsqueeze(0).unsqueeze(0).float(),  # 类别标签                 ], dim=-1))         if len(filtered_boxes) &gt; 0:             results.append(torch.cat(filtered_boxes, dim=0))         else:             results.append([])     return results` 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;
  7W7W7W发表评论（bubbliiiing/detr-pytorch#20）
 
可以试试只保留置信度最高的一个框，utils_bbox.py文件中修改为：
 ***@***.***_grad（）
 def forward（self， outputs， target_sizes， confidence）：
 out_logits， out_bbox = outputs['pred_logits']， outputs['pred_boxes']
     assert len(out_logits) == len(target_sizes)     assert target_sizes.shape[1] == 2     prob = F.softmax(out_logits, -1)     scores, labels = prob[..., :-1].max(-1)     # 转换为 [x0, y0, x1, y1] 格式     boxes = self.box_cxcywh_to_xyxy(out_bbox)     img_h, img_w = target_sizes.unbind(1)     img_h = img_h.float()     img_w = img_w.float()     scale_fct = torch.stack([img_w, img_h, img_w, img_h], dim=1)     boxes = boxes * scale_fct[:, None, :]     results = []     for b_idx in range(len(boxes)):         box_list = boxes[b_idx]         score_list = scores[b_idx]         label_list = labels[b_idx]         unique_labels = torch.unique(label_list)  # 获取所有类别         filtered_boxes = []         for cls in unique_labels:             cls_mask = (label_list == cls) &amp; (score_list &gt; confidence)             cls_boxes = box_list[cls_mask]             cls_scores = score_list[cls_mask]             if len(cls_boxes) &gt; 0:                 # 选择该类别中置信度最高的框                 max_score_idx = torch.argmax(cls_scores)                 filtered_boxes.append(torch.cat([                     cls_boxes[max_score_idx].unsqueeze(0),  # 边界框                     cls_scores[max_score_idx].unsqueeze(0).unsqueeze(0),  # 置信度                     cls.unsqueeze(0).unsqueeze(0).float(),  # 类别标签                 ], dim=-1))         if len(filtered_boxes) &gt; 0:             results.append(torch.cat(filtered_boxes, dim=0))         else:             results.append([])     return results`  
—
直接回复此电子邮件，在 GitHub 上查看，或取消订阅。
您收到此消息是因为您编写了该线程。Message ID: ***@***.***&gt;
